server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/DateHistogramAggregator.java

```diff
@@ -22,35 +22,186 @@
+     * optimized into a {@link FiltersAggregator}. Even when it can't be
+     * optimized, it is going to be marginally faster and consume less memory
+     * than the {@linkplain DateHistogramAggregator} because it doesn't need
+     * to the rounding points and because it can pass precise cardinality
+     * estimates to its child aggregations.
+     */
+    public static Aggregator build(
+        String name,
+        AggregatorFactories factories,
+        Rounding rounding,
+        BucketOrder order,
+        boolean keyed,
+        long minDocCount,
+        @Nullable LongBounds extendedBounds,
+        @Nullable LongBounds hardBounds,
+        ValuesSourceConfig valuesSourceConfig,
+        SearchContext context,
+        Aggregator parent,
+        CardinalityUpperBound cardinality,
+        Map<String, Object> metadata
+    ) throws IOException {
+        Rounding.Prepared preparedRounding = valuesSourceConfig.roundingPreparer().apply(rounding);
+        Aggregator asRange = adaptIntoRangeOrNull(
+            name,
+            factories,
+            rounding,
+            preparedRounding,
+            order,
+            keyed,
+            minDocCount,
+            extendedBounds,
+            hardBounds,
+            valuesSourceConfig,
+            context,
+            parent,
+            cardinality,
+            metadata
+        );
+        if (asRange != null) {
+            return asRange;
+        }
+        return new DateHistogramAggregator(
+            name,
+            factories,
+            rounding,
+            preparedRounding,
+            order,
+            keyed,
+            minDocCount,
+            extendedBounds,
+            hardBounds,
+            valuesSourceConfig,
+            context,
+            parent,
+            cardinality,
+            metadata
+        );
+    }
+
+    private static FromDateRange adaptIntoRangeOrNull(
+        String name,
+        AggregatorFactories factories,
+        Rounding rounding,
+        Rounding.Prepared preparedRounding,
+        BucketOrder order,
+        boolean keyed,
+        long minDocCount,
+        @Nullable LongBounds extendedBounds,
+        @Nullable LongBounds hardBounds,
+        ValuesSourceConfig valuesSourceConfig,
+        SearchContext context,
+        Aggregator parent,
+        CardinalityUpperBound cardinality,
+        Map<String, Object> metadata
+    ) throws IOException {
+        if (hardBounds != null) {
+            return null;
+        }
+        if (valuesSourceConfig.hasValues() == false) {
+            return null;
+        }
+        long[] fixedRoundingPoints = preparedRounding.fixedRoundingPoints();
+        if (fixedRoundingPoints == null) {
+            return null;
+        }
+        // Range aggs use a double to aggregate and we don't want to lose precision.
+        long max = fixedRoundingPoints[fixedRoundingPoints.length - 1];
+        if ((double) max != max) {
```

Comment:
```suggestion
        if ((long) ((double) max) != max) {
```
Let's be explicit about the casting here.  I don't want to have to wonder if some optimization decided to cast the RHS to a double instead of the LHS to a long.  Ditto for other instances of this check.